XenModel/Actions/Pool_Patch/UploadPatchAction.cs

```diff
@@ -117,35 +117,42 @@ private Pool_patch UploadRetailPatch()
             log.InfoFormat("Uploading file '{0}' to server '{1}'", _patchName, master.Name());
             this.Description = string.Format(Messages.UPLOAD_PATCH_UPLOADING_DESCRIPTION, _patchName);
 
-            RelatedTask = Task.create(session, "uploadTask", retailPatchPath);
-
             try
             {
-                var result = HTTPHelper.Put(UpdateProgress, GetCancelling, true, Connection, RelatedTask, ref session, retailPatchPath,
-                    master.address, (HTTP_actions.put_ss)HTTP_actions.put_pool_patch_upload, session.opaque_ref);
-
-                return Connection.WaitForCache(new XenRef<Pool_patch>(result));
+                RelatedTask = Task.create(session, "put_pool_patch_upload_task", master.address);
+                log.DebugFormat("HTTP PUTTING file from {0} to {1}", retailPatchPath, master.address);
+
+                HTTP_actions.put_pool_patch_upload(UpdateProgress,
+                    () => XenAdminConfigManager.Provider.ForcedExiting || GetCancelling(),
+                    XenAdminConfigManager.Provider.GetProxyTimeout(true),
+                    master.address,
+                    XenAdminConfigManager.Provider.GetProxyFromSettings(Connection),
+                    retailPatchPath, RelatedTask.opaque_ref, session.opaque_ref);
+
+                PollToCompletion();
+                return Connection.WaitForCache(new XenRef<Pool_patch>(new XenRef<Pool_patch>(Result)));
             }
-            catch (CancelledException)
+            catch (Exception e)
             {
-                if (deleteFileOnCancel && File.Exists(retailPatchPath))
+                PollToCompletion(suppressFailures: true);
+
+                if (e is TargetInvocationException ex)
                 {
-                    File.Delete(retailPatchPath);
+                    if (ex.InnerException == null)
+                        throw;
+                    else
+                        throw ex.InnerException;
                 }
+
+                if (e is CancelledException || e is HTTP.CancelledException || e.InnerException is CancelledException)
+                {
+                    if (deleteFileOnCancel && File.Exists(retailPatchPath))

```

Comment:
By the looks of https://docs.microsoft.com/en-us/dotnet/api/system.io.file.delete?view=net-5.0#remarks the `File.Exists` check may be skipped.
"If the file to be deleted does not exist, no exception is thrown."